src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -56,16 +100,112 @@ public override async Task<IReadOnlyList<PackageSpec>> GetPackageSpecsAsync(Depe
             return dgSpec;
         }
 
-        public abstract Task<ProjectPackages> GetInstalledAndTransitivePackagesAsync(CancellationToken token);
+        /// <summary>
+        /// Gets the installed (top level) package references for this project. 
+        /// </summary>
+        public override async Task<IEnumerable<PackageReference>> GetInstalledPackagesAsync(CancellationToken token)
+        {
+            ProjectPackages packages = await GetInstalledAndTransitivePackagesAsync(token);
+            return packages.InstalledPackages;
+        }
 
-        private protected IEnumerable<PackageReference> GetPackageReferences(IEnumerable<LibraryDependency> libraries, NuGetFramework targetFramework, Dictionary<string, ProjectInstalledPackage> installedPackages, IList<LockFileTarget> targets)
+        /// <inheritdoc/>
+        public virtual async Task<ProjectPackages> GetInstalledAndTransitivePackagesAsync(CancellationToken token)
+        {
+            PackageSpec packageSpec = null;
+            string assetsPath = null;
+            try
+            {
+                (packageSpec, assetsPath) = await GetCurrentPackageSpecAndAssetsFilePathAsync(token);
+            }
+            catch (ProjectNotNominatedException)
+            {
+            }
+
+            if (packageSpec == null) // null means project is not nominated
+            {
+                IsInstalledAndTransitiveComputationNeeded = true;
+
+                return new ProjectPackages(Array.Empty<PackageReference>(), Array.Empty<TransitivePackageReference>());
+            }
+
+            IList<LockFileTarget> targetsList = null;
+            if (IsInstalledAndTransitiveComputationNeeded)
+            {
+                // clear the transitive packages cache, since we don't know when a dependency has been removed
+                InstalledPackages.Clear();
+                TransitivePackages.Clear();
+                targetsList = await GetTargetsListAsync(assetsPath, token);
+            }
+
+            var frameworkSorter = new NuGetFrameworkSorter();
+
+            // get installed packages
+            List<PackageReference> installedPackages = packageSpec
+                .TargetFrameworks
+                .SelectMany(f => FetchInstalledPackagesList(f.Dependencies, f.FrameworkName, InstalledPackages, targetsList))
+                .GroupBy(p => p.PackageIdentity)
+                .Select(g => g.OrderBy(p => p.TargetFramework, frameworkSorter).First())
+                .ToList();
```

Comment:
Probably no need to call ToList() which does actual calculations and allocation, because later it's still passed to another Linq operation.

Reply:
I will keep the list here to avoid heap allocations. See https://github.com/NuGet/NuGet.Client/pull/4086#issuecomment-855089185